interfaces/cython/cantera/yaml2ck.py

```diff
@@ -0,0 +1,1047 @@
+            )
+
+            # third line has the last two coefficients of the high-temperature range,
+            # first three coefficients of low-temperature range, and "3"
+            text.append(
+                "".join([f"{c:15.8e}" for c in species.thermo.coeffs[6:8]])
+                + "".join([f"{c:15.8e}" for c in species.thermo.coeffs[8:11]])
+                + f"{3:>5}"
+            )
+
+            # fourth and last line has the last four coefficients of the
+            # low-temperature range, and "4"
+            text.append(
+                "".join([f"{c:15.8e}" for c in species.thermo.coeffs[11:15]])
+                + f"{4:>20}\n"
+            )
+
+            thermo_text.append("\n".join(text))
+
+        return thermo_text
+
+    def build_nasa9_text(solution_species):  # TODO: Nonfunctional, general scheme
+        """
+        Would build NASA9-style text string however, Chemkin doesn't accept NASA9
+        currently
+        """
+        # Ex:
+        #   therm NASA9
+        #     200.00  1000.00  6000.00  20000.   3/19/02
+        # e-                Ref-Species. Chase, 1998 3/82.
+        #  3 912/98 E   1.00    0.00    0.00    0.00    0.00 0.000548579903          0.000
+        #     298.150   1000.0007 -2.0 -1.0  0.0  1.0  2.0  3.0  4.0  0.0         6197.428
+        #  0.000000000D+00 0.000000000D+00 2.500000000D+00 0.000000000D+00 0.000000000D+00
+        #  0.000000000D+00 0.000000000D+00                -7.453750000D+02-1.172081224D+01
+        #    1000.000   6000.0007 -2.0 -1.0  0.0  1.0  2.0  3.0  4.0  0.0         6197.428
+        #  0.000000000D+00 0.000000000D+00 2.500000000D+00 0.000000000D+00 0.000000000D+00
+        #  0.000000000D+00 0.000000000D+00                -7.453750000D+02-1.172081224D+01
+        #    6000.000  20000.0007 -2.0 -1.0  0.0  1.0  2.0  3.0  4.0  0.0         6197.428
+        #  0.000000000D+00 0.000000000D+00 2.500000000D+00 0.000000000D+00 0.000000000D+00
+        #  0.000000000D+00 0.000000000D+00                -7.453750000D+02-1.172081224D+01
+        #
+        # Could build NASA9 from NASA7 by setting first 2 coefficients to 0
+        # NASA9 not supported by Chemkin currently though? Comes from NASA CEA mostly
+        pass
+
+    # get list of all thermo models used
+    species_thermo_model = {}
+    for species in solution_species:
+        if species.input_data["thermo"]["model"] != "NASA7":
+            raise NotImplementedError(
+                f"Unsupported thermo model: "
+                f"{species.input_data['thermo']['model']}\n"
+                f"For species: {species.name}"
+            )
+
+    # Get temperature breakpoints
+    temp_range = [1e30, 0]
+    for species in solution_species[0:1]:
+        if species.thermo.min_temp < temp_range[0]:
+            temp_range[0] = species.thermo.min_temp
+
+        if species.thermo.max_temp > temp_range[1]:
+            temp_range[1] = species.thermo.max_temp
```

Comment:
I don't think min and max is the right way to choose these. These are used for default values for species that don't specify the temperatures, so there's no guarantee that those temperature are among the existing set in the file. I think a better solution here is to to what THERMOBUILD from NASA does, whatever that is.

Reply:
This has an error in it from testing. It was supposed to loop through all species and get the maximum and minimum temperatures. The whole mechanism would then be bounded between those. Seems reasonable to me since you'll get errors thrown at you beyond those bounds anyways. Honestly most mechanisms I've seen have it arbitrarily set from 200 to 6000 and it never changes. That's how I had it set initially, but I think you didn't like it.

Reply:
> The whole mechanism would then be bounded between those. Seems reasonable to me since you'll get errors thrown at you beyond those bounds anyways.

I don't think that's what these values do. My (admittedly old) understanding is that these serve as default values when specific values are not specified for a species. Aside from that, I've never gotten any errors (from the thermo, anyways) operating at temperatures outside the limits of the polynomials. The fits are likely garbage, and the simulation most likely won't run, but the thermo itself doesn't throw an error.

> Honestly most mechanisms I've seen have it arbitrarily set from 200 to 6000 and it never changes. That's how I had it set initially, but I think you didn't like it.

Yeah, I don't like anecdata in general when we have to support this feature for everyone who uses Cantera, and who will show up in the issues here or on Google Groups with questions and asking for support 😄 If we can get a canonical reference, such as [ThermoBuild](https://cearun.grc.nasa.gov/ThermoBuild/), I'd feel much better about setting these values as constants.

Reply:
I don't see how it matters what values get put here. They shouldn't end up being used for anything, since this tool will always output species-specific values in the relevant portion of each species entry which should be used instead.

Reply:
Does `ck2yaml` copy the default values into each species if they don't have values set?

Reply:
It would have an error if the T_min, T_mid, and T_max were not set in Cantera. I didn't think it was possible for this to happen?

Reply:
`ck2yaml` (and `ck2cti` before it) will use this as the default value for `T_mid` if it's not specified in the individual species, but requires values for `T_min` and `T_max` (although these bounds are not often checked/enforced in Cantera). Reading the documentation of the CK format that I found, there is no indication that any of these are optional in the species entry, so I don't think the defaults should ever be used in practice.

Reply:
Thanks for checking that @speth! @tsikes this methods of min and max is fine then, I think.

Separately, if it's not used, why is it there! 😠  Ah Chemkin, how I don't miss you at all.